### PR TITLE
Fixes on analytics

### LINKED ISSUE #292: Fixes on analytics
**Is your feature request related to a problem? Please describe.**
Currently, only one service or course can be selected from the filters in analytics, additionally, all courses are shown, even the ones that do not have analytics registered.

**Describe the solution you'd like**
- Select several services and courses in analytics; also, only show the courses that have analytics. 
- Additionally, add a new plot to see analytics per course. 
- Download CSV with the data.

### PR DESCRIPTION
**Pull request by Heldeg**
# Description

 This PR adds new features for analytics plugin:

- Now it allows to select several services and courses to filter the data. Also, Course filter only displays the courses with analytic data.
- It adds a new tab named `Services per course`. This tab shows stacked bar plot about the use of services in each course.
- It adds a new button to download a CSV with all analytic data.
- It adds a new date input to only get data until the a specific date.

Fixes #292

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- With Postman I checked that the inputs don't create errors
-  I checked the behavior of the new tab with different filters

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings

### CODE REVIEW

File: inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+
+def get_course_ids(data):
+    """" Return a list of course ids """
+    return list(map(lambda course: course["course"], data))
+
+
+def create_y_data_list(num_courses, num_services):
+    """ create zero list to represent y data """
+    return [[0] * num_services for i in range(num_courses)]
```
Comment (crdgonzalezca) [inline inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py:15]: I think this function is not necessary. Just do the same logic in the function where you are calling this. You can add a comment that this creates a list of lists containing zeros.

File: inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                        "$push": {
+                            "service": "$_id.service",
+                            "visits": "$visits"
+                        }
+                    }
+                }
+            },
+            {
+                "$project": {
+                    "course": "$_id",
```
Comment (crdgonzalezca) [inline inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py]: `course_id`?

File: inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        input_dict = web.input()
+
+        try:
+            query_parameters = get_api_query_parameters(input_dict)
+        except APIError as error:
+            return error.status_code, {"error": error.return_value}
+
+        data = list(self.get_visits_data(query_parameters))
+
+        course_names = get_course_ids(data)
```
Comment (crdgonzalezca) [inline inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py]: why the variable is named `course_names` and the function returns the ids? Try to have consistency with the names of variables and functions.

**Review by crdgonzalezca — CHANGES_REQUESTED** [review]

File: inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py (hunk 3 above)
Comment (Heldeg) [reply inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py]: yes, in the last moment I decided to change the name of the function `get_course_names` to `get_course_ids` and I forgot to change the names of the variables

File: inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py (hunk 1 above)
Comment (Heldeg) [reply inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py:15]: I think the opposite: It's better to have less comments. Normally the functions change, but not the comments, so it's better if the code is understandable without comments.

File: inginious/frontend/plugins/analytics/utils.py [hunk 4]
```diff
@@ -16,18 +17,41 @@ def get_api_query_parameters(input_dict):
...
+    if start_date or end_date:
+        query_parameters['date'] = {}
+        if start_date:
+            start_date = _convert_string_to_date(start_date)
+            query_parameters['date']['$gte'] = start_date
+        if end_date:
+            end_date = _convert_string_to_date(end_date)
+            query_parameters['date']['$lte'] = end_date
+        if start_date and end_date and (end_date < start_date):
+            raise api.APIError(400, _("start date must be greater than end date"))
```
Comment (crdgonzalezca) [inline inginious/frontend/plugins/analytics/utils.py]:
```suggestion
            raise api.APIError(400, _("The start date must be greater than end date."))
```

**Review by crdgonzalezca — CHANGES_REQUESTED** [review]

File: inginious/frontend/i18n/pt/LC_MESSAGES/messages.po [hunk 5]
```diff
@@ -1307,12 +1307,21 @@ msgid ""
...
 msgstr ""
 
+#: inginious/frontend/plugins/analytics/utils.py:41
+msgid "start date must be greater than end date"
+msgstr ""
+
+#: inginious/frontend/plugins/analytics/utils.py:57
+#, fuzzy
+msgid "Invalid date format"
+msgstr "Formato de e-mail inválido."
```
[the remark this answers is not in the document]
Comment (Heldeg) [reply inginious/frontend/i18n/pt/LC_MESSAGES/messages.po]: Yes, It seems that Babel generates it automatically with similar translations and add `fuzzy` tag to disable it. For that I have problems if the new strings are similars. I haven't done much research, but I think it's due to the version I have.

**Review by crdgonzalezca — CHANGES_REQUESTED** [review]
We are closer, take a look to these couple of comments and also fix the problem with the heat map as commented offline.

File: inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            return error.status_code, {"error": error.return_value}
+
+        data = list(self.get_visits_data(query_parameters))
+
+        course_ids = get_course_ids(data)
+        service_names = self.get_service_names(query_parameters)
+
+        visits = get_visit_values(data, len(course_ids), service_names)
+
+        results = {"x_data": course_ids, "y_data": visits, "services": service_names}
```
[the remark this answers is not in the document]
Comment (Heldeg) [reply inginious/frontend/plugins/analytics/api/stacked_bar_plot_api.py:48]: Yes, and it's matched in `stacked_bar_plot.js`

**Review by crdgonzalezca — APPROVED** [review]

**Pull request merged**